rfcs/010-gateway-refactor/010-gateway-refactor.md

```diff
@@ -0,0 +1,205 @@
+- Start Date: 09-12-2022
+- Status: Proposed
+
+# Gateway refactor
+
+## Summary
+
+This proposes a fulll rewrite of the current gateway service. In addition, it also proposes
+that instead of using the oauth2-proxy in every user session that a centralized
+authentication service fulfils this role for all sessions.
+
+## Motivation
+
+Reasons for rewriting the gateway:
+- the current code base is hard to maintain
+- one service could replace both traefik and the gateway
+- problems with managing sessions across different services/clients
+- using golang should result in more efficient use of resources and improved performance
+
+Reasons for centralized session authentication:
+- running sessions in subdomains which would allow for SSH access
+
+## Design Detail
+
+### Gateway
+
+The gateway has to have sessions with different clients. Some of these are:
+- User servers (i.e. the sidecar container needs be able to swap some kind of ID
+or access token for a Gitlab token)
+- Actual users logging in through the UI
+- Actual users logging in through the CLI
+
+Using JWT access tokens for this purpose is not feasible. This is because access
+tokens from Keycloak expire usually within an hour or several hours. This means
+that anyone who has to remain authenticated for a long time with the gateway
+needs to also get a refresh token and worry about using this to have valid access
+tokens. Accepting an expired access token is not secure.
+
+The gateway should instead have server-side sessions with every "client". For each such
+session the gateway will issue a random, hard to guess ID that will be stored either
+in a cookie or simply as a secret. This is exactly how sessions are handled by the
+ui-server. The cookie or ID is presented with every request to the gateway. The gateway
+looks this ID up in redis and if there is a match then any tokens/credentials stored
+under that key can be added to requests initiated by that ID.
+
+These sessions are managed/started as follows:
+- when a user logs in they get a session
+- when a user server is started the server gets a redis-based session
+- when a login happens through the CLI there is a also a session
+- anonymous users also have redis-based sessions but these are empty
+since there is no additional credentials that need to be managed for anonymous users
+
+The gateway then has the following responsibilities:
+- store refresh and access tokens
+- keep all tokens stored up to date and manage refreshing tokens
+- destroy sessions when certain events occur (i.e. when a logout occurs)
+
+### Centralized Jupyter server authentication
+
+### Diagrams
+
+1. Creating a user server
+
+```mermaid
+graph TD
+    Ingress -- 1<br>Session ID --> Gateway
```

Comment:
I guess `ingress` is the user / client?

Reply:
Yes exactly. 

I had `browser --> ingress --> gateway` everywhere but I removed the browser part to save on space.